cunumeric/deferred.py

```diff
@@ -974,23 +831,92 @@ def contract(
         rhs2_modes,
         mode2extent,
     ):
+        supported_dtypes = [
+            np.float16,
+            np.float32,
+            np.float64,
+            np.complex64,
+            np.complex128,
+        ]
         lhs_thunk = self
 
-        # TODO: More sanity checks (no duplicate modes, no singleton modes, no
-        # broadcasting, ...)
-
-        # Casting should have been handled by the frontend
-        assert lhs_thunk.dtype is rhs1_thunk.dtype
-        assert lhs_thunk.dtype is rhs2_thunk.dtype
+        # Sanity checks
+        # no duplicate modes within an array
+        assert len(lhs_modes) == len(set(lhs_modes))
+        assert len(rhs1_modes) == len(set(rhs1_modes))
+        assert len(rhs2_modes) == len(set(rhs2_modes))
+        # no singleton modes
+        mode_counts = Counter()
+        mode_counts.update(lhs_modes)
+        mode_counts.update(rhs1_modes)
+        mode_counts.update(rhs2_modes)
+        for count in mode_counts.values():
+            assert count == 2 or count == 3
+        # arrays and mode lists agree on dimensionality
+        assert lhs_thunk.ndim == len(lhs_modes)
+        assert rhs1_thunk.ndim == len(rhs1_modes)
+        assert rhs2_thunk.ndim == len(rhs2_modes)
+        # array shapes agree with mode extents (broadcasting should have been
+        # handled by the frontend)
+        assert all(
+            mode2extent[mode] == dim_sz
+            for (mode, dim_sz) in zip(
+                lhs_modes + rhs1_modes + rhs2_modes,
+                lhs_thunk.shape + rhs1_thunk.shape + rhs2_thunk.shape,
+            )
+        )
+        # casting has been handled by the frontend
+        assert lhs_thunk.dtype == rhs1_thunk.dtype
+        assert lhs_thunk.dtype == rhs2_thunk.dtype
 
         # Handle store overlap
         rhs1_thunk = rhs1_thunk._copy_if_overlapping(lhs_thunk)
         rhs2_thunk = rhs2_thunk._copy_if_overlapping(lhs_thunk)
 
+        # Test for special cases where we can use BLAS
+        blas_op = None
+        if any(c != 2 for c in mode_counts.values()):
+            pass
+        elif (
+            len(lhs_modes) == 0
+            and len(rhs1_modes) == 1
+            and len(rhs2_modes) == 1
+        ):
+            # this case works for any arithmetic type, not just floats
+            blas_op = "vv"
```

Comment:
A stylistic suggestion: I guess we could use `IntEnum` here instead.

Reply:
ff85694